### ISSUE TITLE
Adding support for characteristic polynomial of element A over GF

### ISSUE THREAD

**Issue by geostergiop** [mhostetter/galois#220]
As requested, I am opening here a new issue to add support of a function able to calculate the characteristic polynomial of an element A in a GF.

For reference from [wikipedia](https://en.wikipedia.org/wiki/Characteristic_polynomial):
![image](https://user-images.githubusercontent.com/8626213/148788603-9d6cb77f-d2b8-4753-a840-e77361765e6b.png)

Two element matrices may have the same characteristic polynomial but different minimal polynomials, e.g.
![image](https://user-images.githubusercontent.com/8626213/148788891-2817d186-901c-4a6c-abe4-9b8e8e24199d.png)

_Originally posted by @geostergiop in https://github.com/mhostetter/galois/issues/217#issuecomment-1008965049_

**Comment by mhostetter** [comment]
@geostergiop I added this feature in #223. I also added unit tests, so I'm fairly confident in its accuracy. However, can you also be my tester and see if you find any other issues or bugs?  Docs are [here](https://galois.readthedocs.io/en/latest/api/galois.FieldArray.html#galois.FieldArray.characteristic_poly).

Here's an example.

```python
In [1]: import galois                                                                                                            

In [2]: import numpy as np                                                                                                       

In [3]: GF = galois.GF(3**5)                                                                                                     

In [4]: A = GF.Random((3,3)); A                                                                                                  
Out[4]: 
GF([[ 80,  32,   8],
    [115, 145, 102],
    [175,  89,  70]], order=3^5)

In [5]: poly = A.characteristic_poly(); poly                                                                                     
Out[5]: Poly(x^3 + 188x^2 + 25x + 12, GF(3^5))

In [6]: poly.coeffs[-1] == np.linalg.det(-A)                                                                                     
Out[6]: True

In [7]: poly.coeffs[1] == -np.trace(A)                                                                                           
Out[7]: True
```

You'll need to uninstall and reinstall from `master`.

```
$ pip3 uninstall galois
$ pip3 install git+https://github.com/mhostetter/galois.git
```

**Comment by geostergiop** [comment]
Thanks again Matt, as always. Sure, I 'll forward any bug findings I come across, it's a bit further down the road in terms of utilization so you might not get any news on this one for the coming month or so. Just FYI.

Thanks again!
G

**Comment by geostergiop** [comment]
Hi Matt, some application help if you may. It is not clear to me how to invoke `.characteristic_poly()` to create the characteristic polynomial of a given galois.Poly object, since current utilization requires a raw `n x n` matrix input. If I invoke xx on a Binary.Poly, I (obviously) get the `AttributeError: 'BinaryPoly' object has no attribute 'characteristic_poly'` error.

Did you include a direct way to calculate the characteristic poly of a galois.Poly() element?

**Comment by mhostetter** [comment]
Hey, @geostergiop. I'm not sure what you mean by "a characteristic poly of a galois.Poly() element"? I'm also unsure about what you wrote earlier "a function able to calculate the characteristic polynomial of an element A in a GF".

What is a characteristic polynomial of a finite field element? I implemented the characteristic polynomial of a square matrix of finite field elements, like your snippet from Wikipedia.

![image](https://user-images.githubusercontent.com/12112573/150544089-c0572571-a109-4373-94ed-4bcc8503037b.png)

Can you give me an example field, example element, and the correct characteristic polynomial for that element? Or a link to the math/process you desire?

**Comment by geostergiop** [comment]
From Lidl's 2.33 lemma: If T is a linear operator of the finite-dimensional vector space V over the arbitrary field K, then a polynomial is said to annihilate T if `a_n*T^n + ... + a_1*T + a_0*I = 0` where I is the identity operator and 0 the zero operator on V. 

For reference (although I am sure you know these) the uniquely determined monic polynomial of least positive degree with this property is called _minimal polynomial_ (you already support minimal polys) for T. It divides any other polynomial in K[x] annihilating T, and also divides the _characteristic polynomial_ g(x) given by g(x) = det(x*I-T) of degree equal to the dimension of V.

By creating a minimal.Poly(GF(3).vector()) over a GF(2**256), I am interested in testing whether the minimal polynomial and the characteristic polynomial produced over the common vector space from an element x=GF() are the same or not.

**Comment by geostergiop** [comment]
Maybe this can help:
Let L be a finite field extension of K. For every element θ in L define the characteristic polynomial of θ as follows
```
charθ(X) := det (X ⋅ idL − rθ)
where rθ is the K-linear map given by
rθ : L ⟶ L ; x ⟼ xθ
```
Let finally f(X) be the minimal polynomial of θ over K.

Link [here](https://math.stackexchange.com/questions/384206/characteristic-vs-minimal-polynomial)

**Comment by mhostetter** [comment]
Ok, I see. I implemented the characteristic polynomial over a square matrix, but not a single element. Let me take the weekend to work on this and see if I can come up with something.

**Comment by geostergiop** [comment]
I 've been thanking you a lot lately! Thanks Matt. Seems exactly what I'm looking for.
 Will keep you posted in case any bugs come up.

**Comment by mhostetter** [comment]
@geostergiop I just merged #229 onto `master`. So now the minimal polynomials are calculated with `FieldArray.minimal_poly()`, like the characteristic polys, and no longer with `galois.minimal_poly()`.

Here's an example printing the characteristic and minimal polynomials of every field element in `GF(2^4)`. The minimal polynomial always divides the characteristic polynomial.

```python
In [1]: import galois

In [2]: GF = galois.GF(2**4)

In [3]: for a in GF.Elements():
   ...:     print(a)
   ...:     print("\t", a.characteristic_poly())
   ...:     print("\t", a.minimal_poly())
   ...:     print("\t", divmod(a.characteristic_poly(), a.minimal_poly()))
   ...: 
GF(0, order=2^4)
         Poly(x^4, GF(2))
         Poly(x, GF(2))
         (Poly(x^3, GF(2)), Poly(0, GF(2)))
GF(1, order=2^4)
         Poly(x^4 + 1, GF(2))
         Poly(x + 1, GF(2))
         (Poly(x^3 + x^2 + x + 1, GF(2)), Poly(0, GF(2)))
GF(2, order=2^4)
         Poly(x^4 + x + 1, GF(2))
         Poly(x^4 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(3, order=2^4)
         Poly(x^4 + x + 1, GF(2))
         Poly(x^4 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(4, order=2^4)
         Poly(x^4 + x + 1, GF(2))
         Poly(x^4 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(5, order=2^4)
         Poly(x^4 + x + 1, GF(2))
         Poly(x^4 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(6, order=2^4)
         Poly(x^4 + x^2 + 1, GF(2))
         Poly(x^2 + x + 1, GF(2))
         (Poly(x^2 + x + 1, GF(2)), Poly(0, GF(2)))
GF(7, order=2^4)
         Poly(x^4 + x^2 + 1, GF(2))
         Poly(x^2 + x + 1, GF(2))
         (Poly(x^2 + x + 1, GF(2)), Poly(0, GF(2)))
GF(8, order=2^4)
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(9, order=2^4)
         Poly(x^4 + x^3 + 1, GF(2))
         Poly(x^4 + x^3 + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(10, order=2^4)
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(11, order=2^4)
         Poly(x^4 + x^3 + 1, GF(2))
         Poly(x^4 + x^3 + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(12, order=2^4)
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(13, order=2^4)
         Poly(x^4 + x^3 + 1, GF(2))
         Poly(x^4 + x^3 + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(14, order=2^4)
         Poly(x^4 + x^3 + 1, GF(2))
         Poly(x^4 + x^3 + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
GF(15, order=2^4)
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         Poly(x^4 + x^3 + x^2 + x + 1, GF(2))
         (Poly(1, GF(2)), Poly(0, GF(2)))
```

**Comment by geostergiop** [comment]
Thanks for the update, will pull again.
